build-support/scripts/check-allowed-imports.sh

```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env bash
+# Copyright (c) HashiCorp, Inc.
+# SPDX-License-Identifier: MPL-2.0
+
+
+readonly SCRIPT_NAME="$(basename ${BASH_SOURCE[0]})"
+readonly SCRIPT_DIR="$(dirname "${BASH_SOURCE[0]}")"
+readonly SOURCE_DIR="$(dirname "$(dirname "${SCRIPT_DIR}")")"
+readonly FN_DIR="$(dirname "${SCRIPT_DIR}")/functions"
+
+source "${SCRIPT_DIR}/functions.sh"
+
+
+set -uo pipefail
+
+usage() {
+cat <<-EOF
+Usage: ${SCRIPT_NAME} <module root> [<allowed relative package path>...]
+
+Description:
+    Verifies that only the specified packages may be imported from the given module 
+
+Options:
+    -h | --help              Print this help text.
+EOF
+}
+
+function err_usage {
+    err "$1"
+    err ""
+    err "$(usage)"
+}
+
+function main {
+   local module_root=""
+   declare -a allowed_packages=()
+   while test $# -gt 0
+   do
+      case "$1" in
+         -h | --help )
+         usage
+         return 0
+         ;;
+      * )
+         if test -z "$module_root"
+         then
+            module_root="$1"
+         else
+            allowed_packages+="$1"
+         fi
+         shift     
+      esac
+   done
+   
+   check_imports "$module_root" ${allowed_packages[@]+"${allowed_packages[@]}"}
```

Comment:
What does the `allowed_packages[@]+"${allowed_packages[@]}"` incantation do? My bash-fu isn't great!

Reply:
It makes this not error out if no allowed packages are specified (i.e. the whole module is not allowed). Otherwise just `${allowed_packages[@]}` in combination with `set -u` will cause bash to error out due to unbound variables. Even the predelcaration above doesn't prevent this.

Apparently bash 4.2 or something around there fixes this but macOS is still shipping with bash 3.X so I thought I would add the workaround. As for why it works, I have no clue. I just found the workaround, tested it and then dropped it in here.

I suppose I should probably document that though.